src/common/GPUUtils.hpp

```diff
@@ -180,58 +183,92 @@ inline void seq_for(camp::int_seq<T, ts...> const&, Func&& func)
   if (device_ptr_name != host_ptr_name) {                                      \
-    for (int i = 0; i < (length); ++i) {                                       \
-      host_ptr_name[i] = (init_ptr)[i];                                    \
+    for (size_t i = 0; i < static_cast<size_t>(length); ++i) {                 \
+      host_ptr_name[i] = (init_ptr)[i];                                        \
     }                                                                          \
     gpu_type##Errchk( gpu_type##MemcpyAsync( device_ptr_name, host_ptr_name,   \
         (length)*sizeof(device_ptr_name[0]),                                   \
         gpu_type##MemcpyHostToDevice, res.get_stream() ) );                    \
   } else {                                                                     \
-    for (int i = 0; i < (length); ++i) {                                       \
-      device_ptr_name[i] = (init_ptr)[i];                                  \
+    for (size_t i = 0; i < static_cast<size_t>(length); ++i) {                 \
+      device_ptr_name[i] = (init_ptr)[i];                                      \
     }                                                                          \
   }
 
-//
-#define RAJAPERF_GPU_REDUCER_COPY_BACK_IMPL(gpu_type, final_ptr, device_ptr_name, host_ptr_name, length) \
+// Copy back data from device_ptr_name into host_ptr_name
+// if the reduction data space is not host accessible
+#define RAJAPERF_GPU_REDUCER_COPY_BACK_IMPL(gpu_type, device_ptr_name, host_ptr_name, length) \
   if (device_ptr_name != host_ptr_name) {                                      \
     gpu_type##Errchk( gpu_type##MemcpyAsync( host_ptr_name, device_ptr_name,   \
         (length)*sizeof(device_ptr_name[0]),                                   \
         gpu_type##MemcpyDeviceToHost, res.get_stream() ) );                    \
   }                                                                            \
-  gpu_type##Errchk( gpu_type##StreamSynchronize( res.get_stream() ) );         \
-  for (int i = 0; i < (length); ++i) {                                         \
-    (final_ptr)[i] = host_ptr_name[i];                                        \
+  gpu_type##Errchk( gpu_type##StreamSynchronize( res.get_stream() ) );
+
+// Copy data into final_ptr from host_ptr_name
+#define RAJAPERF_GPU_REDUCER_COPY_FINAL_IMPL(final_ptr, host_ptr_name, length) \
+  for (size_t i = 0; i < static_cast<size_t>(length); ++i) {                   \
+    (final_ptr)[i] = host_ptr_name[i];                                         \
   }
 
 
 #define RAJAPERF_CUDA_REDUCER_SETUP(pointer_type, device_ptr_name, host_ptr_name, length) \
   RAJAPERF_GPU_REDUCER_SETUP_IMPL(pointer_type, device_ptr_name, host_ptr_name, length)
-#define RAJAPERF_CUDA_REDUCER_TEARDOWN(device_ptr_name, host_ptr_name)          \
+#define RAJAPERF_CUDA_REDUCER_TEARDOWN(device_ptr_name, host_ptr_name) \
```

Comment:
Would it be better to have one set of macros and pass 'cuda' or 'hip' (or whatever we may need in the future) as an arg? It looks like the macros invoked by the `_CUDA_` and `_HIP_` outer macros do that.

Reply:
I'm not sure that these macros could be expanded to anything else in the future because this relies on the similarity between cuda and hip. I made the outer macros so that the implementations could differ more if necessary though they don't in their current form.
There are some of these macros that are the same for the `_CUDA_` and `_HIP_` versions that are the same either way. Would it be better to just use the base macro in that case?

Reply:
I understand. I'm OK with it as is. I think usage is clearer if usage is similar for all cases; i.e., I don't think having some macros without cuda or hip in their name and some with is a good idea -- too confusing.